File: packages/klarna-network-core/android/src/main/java/com/klarna/mobile/sdk/reactnative/klarnanetwork/core/KlarnaNetworkCoreModuleImpl.kt
package com.klarna.mobile.sdk.reactnative.klarnanetwork.core

import android.util.Log
import androidx.annotation.RestrictTo
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.klarna.mobile.sdk.klarna.network.core.api.common.KlarnaResult
import com.klarna.mobile.sdk.klarna.network.core.api.common.KlarnaSDKError
import com.klarna.mobile.sdk.klarna.network.core.api.klarna.Klarna
import com.klarna.mobile.sdk.klarna.network.core.api.klarna.KlarnaConfiguration
import com.klarna.mobile.sdk.klarna.network.core.api.metadata.KlarnaIntegrationMetadata
import com.klarna.mobile.sdk.klarna.network.core.api.metadata.KlarnaIntegratorMetadata
import com.klarna.mobile.sdk.klarna.network.core.api.metadata.KlarnaOriginatorMetadata
import java.util.concurrent.ConcurrentHashMap

class KlarnaNetworkCoreModuleImpl(
    private val reactContext: ReactApplicationContext,
    private val klarnaFactory: (ReactApplicationContext, KlarnaConfiguration) -> KlarnaResult<Klarna, KlarnaSDKError> = { ctx, cfg ->
        Klarna.initialize(ctx, cfg)
    }
) {

    companion object {
        const val NAME = "KlarnaNetworkCore"
        private const val KEY_CLIENT_ID = "clientId"
        private const val KEY_ACCOUNT_ID = "accountId"
        private const val KEY_LOCALE = "locale"
        private const val KEY_KLARNA_NETWORK_SESSION_TOKEN = "klarnaNetworkSessionToken"
        private const val KEY_INTEGRATOR = "integrator"
        private const val KEY_ORIGINATORS = "originators"
        private const val KEY_NAME = "name"
        private const val KEY_SESSION_REFERENCE = "sessionReference"
        private const val KEY_MODULE_NAME = "moduleName"
        private const val KEY_MODULE_VERSION = "moduleVersion"
        private const val ERROR_INSTANCE_NOT_FOUND = "No instance found for the given instanceId. Call initialize first."

        // Sibling network packages (messaging, payment, …) look up the
        // Klarna instance by instanceId received over the JS bridge.
        // Companion-static so the lookup is reachable without an instance of
        // the module class. The visibility is broader than ideal — it should
        // be migrated to a @RequiresOptIn-gated API alongside the iOS @_spi
        // tag once that work lands on core.
        @JvmStatic
        val instances = ConcurrentHashMap<String, Klarna>()

        @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP_PREFIX)
        fun getInstance(instanceId: String): Klarna? = instances[instanceId]
    }

    fun initialize(instanceId: String, configuration: ReadableMap?, promise: Promise?) {
        val clientId = configuration?.getString(KEY_CLIENT_ID) ?: run {
            promise?.reject(NAME, "clientId is required")
            return
        }
        // appReturnUrl from KlarnaConfiguration type is ignored on android side since is not needed.
        val config = KlarnaConfiguration(
            clientId = clientId,
            accountId = configuration.getString(KEY_ACCOUNT_ID),
            locale = configuration.getString(KEY_LOCALE),
            klarnaNetworkSessionToken = configuration.getString(KEY_KLARNA_NETWORK_SESSION_TOKEN)
        )
        when (val result = klarnaFactory(reactContext, config)) {
            is KlarnaResult.Success -> {
                instances[instanceId] = result.value
                promise?.resolve(null)
            }
            is KlarnaResult.Failure -> {
                promise?.reject(result.error.name, result.error.message)
            }
        }
    }

    fun getSessionToken(instanceId: String, promise: Promise?) {
        val klarna = instances[instanceId] ?: run {
            promise?.reject(NAME, ERROR_INSTANCE_NOT_FOUND)
            return
        }
        klarna.network.session.token { result ->
            when (result) {
                is KlarnaResult.Success -> promise?.resolve(result.value)
                is KlarnaResult.Failure -> promise?.reject(result.error.name, result.error.message)
            }
        }
    }

    fun clearSession(instanceId: String, promise: Promise?) {
        val klarna = instances[instanceId] ?: run {
            promise?.reject(NAME, ERROR_INSTANCE_NOT_FOUND)
            return
        }
        klarna.network.session.clear { result ->
            when (result) {
                is KlarnaResult.Success -> promise?.resolve(null)
                is KlarnaResult.Failure -> promise?.reject(result.error.name, result.error.message)
            }
        }
    }

    fun handleReturnUrl(url: String?, promise: Promise?) {
        Log.d(NAME, "handleReturnUrl is not supported on Android.")
        promise?.resolve(false)
    }

    fun setIntegrationMetadata(instanceId: String, metadata: ReadableMap?) {
        val klarna = instances[instanceId] ?: run {
            Log.d(NAME, "setIntegrationMetadata: no instance found for instanceId: $instanceId")
            return
        }
        val integratorMap = metadata?.getMap(KEY_INTEGRATOR) ?: run {
            Log.d(NAME, "setIntegrationMetadata: integrator map is missing")
            return
        }
        val integrator = KlarnaIntegratorMetadata(
            name = integratorMap.getString(KEY_NAME) ?: run {
                Log.d(NAME, "setIntegrationMetadata: integrator name is missing")
                return
            },
            sessionReference = integratorMap.getString(KEY_SESSION_REFERENCE) ?: run {
                Log.d(NAME, "setIntegrationMetadata: integrator sessionReference is missing")
                return
            },
            moduleName = integratorMap.getString(KEY_MODULE_NAME),
            moduleVersion = integratorMap.getString(KEY_MODULE_VERSION)
        )
        val originatorsArray: ReadableArray? = metadata.getArray(KEY_ORIGINATORS)
        val originators = originatorsArray?.let { arr ->
            (0 until arr.size()).mapNotNull { i ->
                arr.getMap(i)?.let { origMap ->
                    KlarnaOriginatorMetadata(
                        name = origMap.getString(KEY_NAME) ?: return@mapNotNull null,
                        sessionReference = origMap.getString(KEY_SESSION_REFERENCE) ?: return@mapNotNull null,
                        moduleName = origMap.getString(KEY_MODULE_NAME),
                        moduleVersion = origMap.getString(KEY_MODULE_VERSION)
                    )
                }
            }
        }
        klarna.integrationMetadata = KlarnaIntegrationMetadata(integrator, originators)
    }

    fun dispose(instanceId: String, promise: Promise?) {
        instances.remove(instanceId)
        promise?.resolve(null)
    }
}

File: packages/klarna-network-core/ios/Sources/KlarnaNetworkCoreModuleImpl.swift
import Foundation
import KlarnaNetworkCore

// MARK: - KlarnaNetworkCoreImplError

struct KlarnaNetworkCoreImplError: Error {
    let name: String
    let message: String
}

// MARK: - KlarnaNetworkCoreInstance

protocol KlarnaNetworkCoreInstance: AnyObject {
    func sessionToken(onSuccess: @escaping (String) -> Void, onFailure: @escaping (KlarnaNetworkCoreImplError) -> Void)
    func clearSession(onSuccess: @escaping () -> Void, onFailure: @escaping (KlarnaNetworkCoreImplError) -> Void)
    var integrationMetadata: KlarnaIntegrationMetadata? { get set }
}

// MARK: - Klarna + KlarnaNetworkCoreInstance

extension Klarna: KlarnaNetworkCoreInstance {
    func sessionToken(
        onSuccess: @escaping (String) -> Void,
        onFailure: @escaping (KlarnaNetworkCoreImplError) -> Void
    ) {
        network.session.token { result in
            switch result {
            case .success(let token): onSuccess(token)
            case .failure(let error): onFailure(KlarnaNetworkCoreImplError(name: error.name, message: error.message))
            }
        }
    }

    func clearSession(
        onSuccess: @escaping () -> Void,
        onFailure: @escaping (KlarnaNetworkCoreImplError) -> Void
    ) {
        network.session.clear { result in
            switch result {
            case .success: onSuccess()
            case .failure(let error): onFailure(KlarnaNetworkCoreImplError(name: error.name, message: error.message))
            }
        }
    }
}

// MARK: - InstanceStore

// Sibling network packages read this map from arbitrary threads via
// `getInstance(instanceId:)`, while initialize/dispose write from the JS
// thread. A serial queue serializes all access so concurrent reads can't
// observe a torn dictionary or race with a write.
final class InstanceStore {
    private let queue = DispatchQueue(label: "com.klarna.rnnetworkcore.instances")
    private var storage: [String: any KlarnaNetworkCoreInstance] = [:]

    subscript(key: String) -> (any KlarnaNetworkCoreInstance)? {
        get { queue.sync { storage[key] } }
        set { queue.sync { storage[key] = newValue } }
    }

    @discardableResult
    func removeValue(forKey key: String) -> (any KlarnaNetworkCoreInstance)? {
        queue.sync { storage.removeValue(forKey: key) }
    }

    func removeAll() {
        queue.sync { storage.removeAll() }
    }
}

// MARK: - KlarnaNetworkCoreModuleImpl

@objc(KlarnaNetworkCoreModuleImpl)
public class KlarnaNetworkCoreModuleImpl: NSObject {

    private static let moduleName = "KlarnaNetworkCore"
    private static let errorInstanceNotFound = "No instance found for the given instanceId. Call initialize first."
    private static let keyName = "name"
    private static let keySessionReference = "sessionReference"
    private static let keyModuleName = "moduleName"
    private static let keyModuleVersion = "moduleVersion"

    static let instances = InstanceStore()

    private let klarnaFactory: (KlarnaConfiguration) -> Result<any KlarnaNetworkCoreInstance, KlarnaNetworkCoreImplError>

    public override init() {
        self.klarnaFactory = { configuration in
            Klarna.initialize(configuration: configuration)
                .map { $0 as any KlarnaNetworkCoreInstance }
                .mapError { KlarnaNetworkCoreImplError(name: $0.name, message: $0.message) }
        }
        super.init()
    }

    init(klarnaFactory: @escaping (KlarnaConfiguration) -> Result<any KlarnaNetworkCoreInstance, KlarnaNetworkCoreImplError>) {
        self.klarnaFactory = klarnaFactory
        super.init()
    }

    @_spi(RNKlarnaNetworkCore)
    public static func getInstance(instanceId: String) -> Klarna? {
        instances[instanceId] as? Klarna
    }

    @objc public func initialize(
        instanceId: String,
        clientId: String,
        appReturnUrl: String,
        accountId: String?,
        locale: String?,
        klarnaNetworkSessionToken: String?,
        resolve: @escaping (Any?) -> Void,
        reject: @escaping (String?, String?, Error?) -> Void
    ) {
        let configuration = KlarnaConfiguration(
            accountId: accountId,
            clientId: clientId,
            locale: locale,
            appReturnUrl: appReturnUrl,
            klarnaNetworkSessionToken: klarnaNetworkSessionToken
        )
        let result = klarnaFactory(configuration)
        switch result {
        case .success(let instance):
            Self.instances[instanceId] = instance
            resolve(nil)
        case .failure(let error):
            reject(error.name, error.message, nil)
        }
    }

    @objc public func getSessionToken(
        instanceId: String,
        resolve: @escaping (Any?) -> Void,
        reject: @escaping (String?, String?, Error?) -> Void
    ) {
        guard let instance = Self.instances[instanceId] else {
            reject(Self.moduleName, Self.errorInstanceNotFound, nil)
            return
        }
        instance.sessionToken(
            onSuccess: { token in resolve(token) },
            onFailure: { error in reject(error.name, error.message, nil) }
        )
    }

    @objc public func clearSession(
        instanceId: String,
        resolve: @escaping (Any?) -> Void,
        reject: @escaping (String?, String?, Error?) -> Void
    ) {
        guard let instance = Self.instances[instanceId] else {
            reject(Self.moduleName, Self.errorInstanceNotFound, nil)
            return
        }
        instance.clearSession(
            onSuccess: { resolve(nil) },
            onFailure: { error in reject(error.name, error.message, nil) }
        )
    }

    @objc public func handleReturnUrl(
        urlString: String,
        resolve: @escaping (Any?) -> Void,
        reject: @escaping (String?, String?, Error?) -> Void
    ) {
        guard let url = URL(string: urlString) else {
            resolve(false)
            return
        }
        DispatchQueue.main.async {
            resolve(Klarna.handleReturnUrl(url: url))
        }
    }

    @objc public func setIntegrationMetadata(
        instanceId: String,
        integratorName: String,
        integratorSessionReference: String,
        integratorModuleName: String?,
        integratorModuleVersion: String?,
        originators: NSArray?
    ) {
        guard let instance = Self.instances[instanceId] else {
            return
        }
        let integrator = KlarnaIntegratorMetadata(
            name: integratorName,
            sessionReference: integratorSessionReference,
            moduleName: integratorModuleName,
            moduleVersion: integratorModuleVersion
        )
        let originatorsList: [KlarnaOriginatorMetadata]? = originators?.compactMap { item in
            guard let dict = item as? [String: String],
                  let name = dict[Self.keyName],
                  let sessionRef = dict[Self.keySessionReference] else { return nil }
            return KlarnaOriginatorMetadata(
                name: name,
                sessionReference: sessionRef,
                moduleName: dict[Self.keyModuleName],
                moduleVersion: dict[Self.keyModuleVersion]
            )
        }
        instance.integrationMetadata = KlarnaIntegrationMetadata(
            integrator: integrator,
            originators: originatorsList
        )
    }

    @objc public func dispose(
        instanceId: String,
        resolve: @escaping (Any?) -> Void,
        reject: @escaping (String?, String?, Error?) -> Void
    ) {
        Self.instances.removeValue(forKey: instanceId)
        resolve(nil)
    }
}
